Consolidate accounts onboarding: Claim Google Ads account via the combo card

ankitrox opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Closes #2582 

_Replace this with a good description of your changes & reasoning._


### Screenshots:

<img width="1097" alt="Screenshot 2024-10-16 at 8 25 44 PM" src="https://github.com/user-attachments/assets/f085cbbc-23a2-4f59-bd82-7eaf452f93e8">

<img width="1082" alt="Screenshot 2024-10-16 at 8 24 53 PM" src="https://github.com/user-attachments/assets/9d8b764f-8c18-49ec-9e25-02ea192f8810">

### Detailed test instructions:

1. Once the MC and Ads accounts are created for the user, you should be able to see the ads accounts claim card.
2. Click on the `Claim your Google Ads account` button, it will open the pop-up window to accept the invitation.
3. If the invitation has been accepted and pop-up is closed, there would be a `Updating...` spinner in place of `Claim your Google Ads account` button and eventually claim card will go away and in place of that green colored notice will be shown as shown in screenshot.
4. If the invitation has not been accepted, claim card will stay.

### Additional details:

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation

If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.

Add the `changelog: none` label if no changelog entry is needed.
-->
### Changelog entry

ankitrox committed:
Claim New Ads Accounts in the Google Combo Accounts Card.

ankitrox committed:
Basic function of the claim card.

ankitrox committed:
Display conversion action notice.

ankitrox committed:
Setup conversion action once invite accepted.

ankitrox committed:
Trigger simgle request for conversion action.

ankitrox committed:
code comment.

ankitrox committed:
Add E2E tests.

ankitrox committed:
Remove test code.

ankitrox committed:
Remove test code.

ankitrox committed:
Fix css lint.

asvinb committed:
Set width for claim card.

ankitrox committed:
Resolve conflicts and merge base branch.

asvinb commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account-button.js:
```diff
@@ -0,0 +1,73 @@
+ * @param {boolean} [props.loading] Whether the button is in a loading state.
+ * @param {Object} props.restProps Props to be forwarded to AppButton.
+ */
+const ClaimAdsAccountButton = ( {
```
@ankitrox Any idea why we cannot use the `ClaimAccountButton` from `js/src/components/google-ads-account-card/claim-account-button.js`?

ankitrox replied:
Updated existing `js/src/components/google-ads-account-card/claim-account-button.js` to accommodate `loading` prop.

asvinb commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account-button.js:
```diff
@@ -0,0 +1,73 @@
+		onClick( event );
+	};
+
+	if ( loading ) {
```
@ankitrox I think we should be able to do the same thing by passing the `loading` prop to `AppButton` along with the `text` prop.

asvinb commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js:
```diff
@@ -0,0 +1,114 @@
+ */
+const ClaimAdsAccount = () => {
+	const [ , forceUpdate ] = useState( 0 );
+	const isWindowFocused = useWindowFocus();
```
@ankitrox Can we use the same logic in `js/src/components/google-ads-account-card/claim-account/index.js`
```js
const { fetchGoogleAdsAccountStatus } = useAppDispatch();
useWindowFocusCallbackIntervalEffect( fetchGoogleAdsAccountStatus, 30 );
```

asvinb commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js:
```diff
@@ -0,0 +1,114 @@
+
+	useEffect( () => {
+		if ( isWindowFocused && claimButtonClickedRef.current ) {
+			invalidateResolution( 'getGoogleAdsAccountStatus' );
```
@ankitrox There won't be the need to invalidate the data when we'll use the suggested approach above, i.e using `useWindowFocusCallbackIntervalEffect`

asvinb commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js:
```diff
@@ -0,0 +1,114 @@
+			step === 'conversion_action' &&
+			! claimInProgressRef.current
+		) {
+			claimInProgressRef.current = true;
```
@ankitrox I don't think we should check whether there's a claim in progress to upsert an Ads account.  I think `hasAccess` and `step === 'conversion_action'` should be enough:
useEffect( () => {
		if ( hasAccess === true && step === 'conversion_action' ) {
			upsertAdsAccount();
		}
	}, [ hasAccess, upsertAdsAccount, step ] );

ankitrox replied:
This has been changed to

```js
useEffect( () => {
if ( hasAccess === true && step === 'conversion_action' ) {
	upsertAdsAccount();
}
}, [ hasAccess, step, upsertAdsAccount ] );
```

asvinb commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js:
```diff
@@ -0,0 +1,114 @@
+	 * 1. The user has access to the Google Ads account.
+	 * 2. The resolution has not finished and the claim button has not been clicked.
+	 */
+	if (
```
@ankitrox I think the condition to render `ClaimAdsAccount` should be in `ConnectedGoogleComboAccountCard`
and we can check as follows:
```js
const shouldClaimGoogleAdsAccount = Boolean(
		googleAdsAccount.id && hasAccess === false
	);
```

Thus we won't need to track for any button click. Side note, we don't need to check for `! hasFinishedResolution` since `hasAccess` will be falsy if that's not the case.

asvinb commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js:
```diff
@@ -0,0 +1,114 @@
+						<ClaimAdsAccountButton
+							inviteLink={ inviteLink }
+							loading={ isLoading }
+							onClick={ () => {
```
@ankitrox The only thing we need from `onClick` is to have an internal state for the `loading` feature. We don't need to track whether it was clicked or not.

asvinb commented on js/src/components/google-combo-account-card/connected-google-combo-account-card.js:
```diff
@@ -93,19 +97,31 @@ const ConnectedGoogleComboAccountCard = () => {
 
+	const getCardDescription = () => {
+		return (
+			<>
```
@ankitrox Unnecessary fragment.
```suggestion

```

asvinb commented on js/src/components/google-combo-account-card/connected-google-combo-account-card.js:
```diff
@@ -93,19 +97,31 @@ const ConnectedGoogleComboAccountCard = () => {
-					isCreatingWhichAccount={ isCreatingWhichAccount }
-				/>
-			}
+			description={ getCardDescription() }
```
@ankitrox Can we not render `AccountCreationDescription` directly? Why a function? 🤔 
```js
<AccountCreationDescription
	isCreatingWhichAccount={ isCreatingWhichAccount }
/>
```

asvinb commented on js/src/components/google-combo-account-card/connected-google-combo-account-card.js:
```diff
@@ -93,19 +97,31 @@ const ConnectedGoogleComboAccountCard = () => {
-		/>
+		>
+			{ showSuccessNotice && <ConversionMeasurementNotice /> }
+			<ClaimAdsAccount />
```
@ankitrox Check my previous comments. Let's add the condition to display `ClaimAdsAccount`

ankitrox committed:
Resolve conflicts and merge base branch.

ankitrox committed:
Resolve conflicts and merge base branch.

ankitrox committed:
Resolve conflicts and merge base branch.

ankitrox committed:
CR feedback.

ankitrox committed:
Arrange module imports.

joemcgill commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js:
```diff
@@ -0,0 +1,99 @@
+	const { hasAccess, step } = useGoogleAdsAccountStatus();
+	const [ upsertAdsAccount ] = useUpsertAdsAccount();
+
+	const checkUpdatedAdsAccountStatus = useCallback( async () => {
+		setUpdating( true );
+		await fetchGoogleAdsAccountStatus();
+		setUpdating( false );
+	}, [ fetchGoogleAdsAccountStatus ] );
+
+	useWindowFocusCallbackIntervalEffect( checkUpdatedAdsAccountStatus, 30 );
```
This looks very close. One thing that I think we can do to improve this is to do something similar to what the `useAutoCheckBillingStatusEffect()` hook [does here](https://github.com/woocommerce/google-listings-and-ads/blob/93eeba40ca59a10fd200473c558758623c7629ed/js/src/components/paid-ads/billing-card/useAutoCheckBillingStatusEffect.js#L43-L50) and return early from this callback once we no longer need to check on the claim status of the account.

Here's a quick example:


```diff
diff --git a/js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js b/js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js
index 547ef9927..fb22d1f02 100644
--- a/js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js
+++ b/js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js
@@ -29,11 +29,19 @@ const ClaimAdsAccount = () => {
 	const { hasAccess, step } = useGoogleAdsAccountStatus();
 	const [ upsertAdsAccount ] = useUpsertAdsAccount();
 
+	const shouldClaimGoogleAdsAccount = Boolean(
+		googleAdsAccount.id && hasAccess === false
+	);
+
 	const checkUpdatedAdsAccountStatus = useCallback( async () => {
+		if ( ! shouldClaimGoogleAdsAccount ) {
+			return;
+		}
+
 		setUpdating( true );
 		await fetchGoogleAdsAccountStatus();
 		setUpdating( false );
-	}, [ fetchGoogleAdsAccountStatus ] );
+	}, [ fetchGoogleAdsAccountStatus, hasAccess ] );
 
 	useWindowFocusCallbackIntervalEffect( checkUpdatedAdsAccountStatus, 30 );
 
@@ -43,9 +51,7 @@ const ClaimAdsAccount = () => {
 		}
 	}, [ hasAccess, step, upsertAdsAccount ] );
 
-	const shouldClaimGoogleAdsAccount = Boolean(
-		googleAdsAccount.id && hasAccess === false
-	);
+
 
 	if ( ! shouldClaimGoogleAdsAccount ) {
 		return null;
```

joemcgill commented on js/src/components/google-combo-account-card/connected-google-combo-account-card.js:
```diff
@@ -22,14 +24,21 @@ const ConnectedGoogleComboAccountCard = () => {
+				}
+			>
+				<ConversionMeasurementNotice />
+				<ClaimAdsAccount />
```
An alternative way we could avoid needing to poll the `ads/account-status` API is to move the `shouldClaimGoogleAdsAccount` logic to this component, and only render this component if `shouldClaimGoogleAdsAccount` is `true`.

ankitrox replied:
Changed it as per [this comment](https://github.com/woocommerce/google-listings-and-ads/pull/2644#issuecomment-2436874661)

ankitrox committed:
Merge branch 'feature/2567-kickoff-mc-ads-account-creation' into update/2582-claim-ads-account

ankitrox committed:
Add ConnectedAccountsActions component.

ankitrox commented:
Thanks for reviewing it @joemcgill.

I have added `ConnectedAccountsActions` component which will be displayed with `helper` prop of the `AccountCard`. This will enable us to simplify the styling as can be seen in [this change](https://github.com/woocommerce/google-listings-and-ads/pull/2644/commits/bb57a8c0247b5aab12ad8439bf899e3f4bff001b#diff-f416ef56fa86302ac2c95f8583f3bf81b0953275fd3ff121f85eb3b5cb56807eL4-L10).

Also, this component will hold the business logic like whether or not to show the other components like claim ads account or conversion measurement notice so that the other components like `ClaimAdsAccount` and `ConversionMeasurementNotice` will be more UI focused which enables them to be easily testable.

Over to you for another review.

joemcgill commented on js/src/components/google-combo-account-card/connected-google-combo-account-card.js:
```diff
@@ -24,21 +23,14 @@ const ConnectedGoogleComboAccountCard = () => {
+			alignIcon="top"
+			className="gla-google-combo-account-card--connected"
+			description={ text || <AccountDetails /> }
+			helper={ subText || <ConnectedAccountsActions /> }
```
Using the `helper` prop for this purpose is not ideal, as that is usually meant to be used for the italicized text shown on cards. Instead, can we pass the `ConnectedAccountsActions` to the body of the card, like you were doing previously?

Doing so should also let you avoid needing to override the color and font-style CSS for this component.

joemcgill commented on js/src/components/google-combo-account-card/connected-google-combo-account-card.scss:
```diff
@@ -3,4 +3,18 @@
+	.gla-connected-accounts-actions {
+		color: $black;
+		font-style: normal;
+		margin-top: 10px;
```
Prefer `$grid-unit-10;`

ankitrox replied:
This is no longer required.

joemcgill commented on js/src/components/google-combo-account-card/connected-google-combo-account-card.scss:
```diff
@@ -3,4 +3,18 @@
 	}
+
+	.gla-connected-accounts-actions {
+		color: $black;
+		font-style: normal;
```
Likely unnecessary if this component is rendered in the Card body instead of where the helper text is rendered.

ankitrox replied:
This is no longer required.

joemcgill commented on js/src/components/google-combo-account-card/connected-accounts-actions.js:
```diff
@@ -0,0 +1,31 @@
+		googleAdsAccount.id && hasAccess === false
+	);
+
+	return (
+		<div className="gla-connected-accounts-actions">
+			{ shouldClaimGoogleAdsAccount && <ClaimAdsAccount /> }
+			{ isReady && <ConversionMeasurementNotice /> }
+		</div>
+	);
```
I like this separation of concerns.

ankitrox replied:
Thanks.

joemcgill commented on tests/e2e/specs/setup-mc/step-1-accounts.test.js:
```diff
@@ -357,6 +357,96 @@ test.describe( 'Set up accounts', () => {
 			} );
 		} );
+
+		test.describe( 'Claim Google Ads Account adsaccountclaim', () => {
```
The description looks to have a typo. Also, I think it would be better to move this set of tests out of the 'Connect Google account' set of tests since they seem to be a self-contained set of tests for confirming the claim behavior. I would suggest a section that follows after that is focused  entirely on the claim flow. Ex:

```
test.describe( 'Connect Google account', () => {
  // ...test logic
};

test.describe( 'Claim Google Ads account', () => {
  test.beforeEach( async () => {
    // Mock the ads account in a claim state.
  } );

  test( 'should see the claim button', async () => {}

  test( 'should see... (etc)', async () => {}
};
```

joemcgill commented on tests/e2e/specs/setup-mc/step-1-accounts.test.js:
```diff
@@ -357,6 +357,96 @@ test.describe( 'Set up accounts', () => {
+					name: 'gla',
+				} );
+
+				await setUpAccountsPage.fulfillAdsAccountStatus( {
+					has_access: false,
+					invite_link: 'https://example.com',
+					step: 'account_access',
+				} );
```
We have a mock for this.

```suggestion
				await setUpAccountsPage.mockAdsStatusNotClaimed();
```

joemcgill commented on tests/e2e/specs/setup-mc/step-1-accounts.test.js:
```diff
@@ -357,6 +357,96 @@ test.describe( 'Set up accounts', () => {
+				await setUpAccountsPage.mockGoogleConnected();
+				await setUpAccountsPage.mockMCConnected();
+
+				await setUpAccountsPage.fulfillAdsConnection( {
+					id: 12345,
+					currency: 'USD',
+					status: 'incomplete',
+					step: 'account_access',
+					sub_account: true,
+					symbol: '$',
+				} );
```
Let's turn this into a mock, similar to `mockAdsAccountIncomplete()` ([ref](https://github.com/woocommerce/google-listings-and-ads/blob/599c2d209707ce06f382fd82e0c181bd04543e46/tests/e2e/utils/mock-requests.js#L577-L590)).

```js
	/**
	 * Mock Google Ads account as connected but unclaimed.
	 *
	 * @return {Promise<void>}
	 */
	async mockAdsAccountUnclaimed( id = 12345 ) {
		await this.fulfillAdsConnection( {
			id,
			currency: 'TWD',
			symbol: 'NT$',
			status: 'incomplete',
			step: 'billing',
		} );
	}
```

joemcgill commented on tests/e2e/specs/setup-mc/step-1-accounts.test.js:
```diff
@@ -357,6 +357,96 @@ test.describe( 'Set up accounts', () => {
 		} );
+
+		test.describe( 'Claim Google Ads Account adsaccountclaim', () => {
+			test.beforeEach( async () => {
```
If we can run these tests sequentially without needing to set up and refreshing the page each time it will speed up the tests. For more, see [these docs](https://playwright.dev/docs/api/class-test#test-before-all).

```suggestion
			test.beforeAll( async () => {
```

ankitrox committed:
Merge branch 'feature/2567-kickoff-mc-ads-account-creation' into update/2582-claim-ads-account

joemcgill commented on tests/e2e/specs/setup-mc/step-1-accounts.test.js:
```diff
@@ -277,6 +277,96 @@ test.describe( 'Set up accounts', () => {
+			} );
+
+			test.describe( 'Account claimed', () => {
+				test.beforeEach( async () => {
+					await setUpAccountsPage.fulfillAdsAccountStatus( {
+						has_access: true,
+						invite_link: '',
+						step: '',
+					} );
+
+					await setUpAccountsPage.fulfillAdsConnection( {
+						id: 78787878,
+						currency: 'USD',
+						status: 'connected',
+						symbol: '$',
+					} );
+
+					await setUpAccountsPage.goto();
+				} );
```
Could this be done from the previous test by triggering a focus event rather than refreshing the page (e.g. `await setUpAccountsPage.goto();`)? This should also use `test.beforeAll` instead of `test.beforeEach`.

ankitrox replied:
The focus event in playwright does not work as expected, that's why this test has been written separately like this.

**References:**
https://github.com/microsoft/playwright/issues/3570
https://github.com/microsoft/playwright/issues/15409

joemcgill reviewed (commented):
Nice one. A few more suggestions.

ankitrox committed:
Merge branch 'feature/2567-kickoff-mc-ads-account-creation' into update/2582-claim-ads-account

ankitrox committed:
E2E tests improvements.

ankitrox commented:
@joemcgill I've addressed the feedback and responded to some comments. Over to you for another pass.

joemcgill reviewed (commented):
This looks very close. `ClaimAdsAccount` component returns a `Section.Card.Body` component, which takes care of a lot of the padding. Currently this is looking a bit off but I think simplifying it to this should work:

```diff
diff --git a/js/src/components/google-combo-account-card/connected-google-combo-account-card.scss b/js/src/components/google-combo-account-card/connected-google-combo-account-card.scss
index 4578751e0..c673e91b2 100644
--- a/js/src/components/google-combo-account-card/connected-google-combo-account-card.scss
+++ b/js/src/components/google-combo-account-card/connected-google-combo-account-card.scss
@@ -6,14 +6,8 @@
 
 	.gla-connected-accounts-actions {
 		@media (min-width: $break-small) {
-			margin-left: auto;
-
 			// 40px is the icon size and $grid-unit-20 is 16px for the gap between flex items.
-			width: calc(100% - 40px - #{$grid-unit-20});
+			padding-left: calc(40px + $grid-unit-20);
 		}
 	}
-
-	&:has(.gla-connected-accounts-actions) .components-card-body:last-of-type {
-		padding: $grid-unit-10;
-	}
 }
```

**Before**
<img width="677" alt="image" src="https://github.com/user-attachments/assets/87036bc5-f3f2-4393-8095-bcdfa70ec5c4">

**After**
<img width="688" alt="Screenshot 2024-10-26 at 4 41 54 PM" src="https://github.com/user-attachments/assets/ed55f187-b452-46b4-b785-26eca126e84e">

ankitrox committed:
Merge branch 'feature/2567-kickoff-mc-ads-account-creation' into update/2582-claim-ads-account

ankitrox committed:
Remove redundant css rules.

ankitrox commented:
Thank you @joemcgill . I've made the changes you requested. Assigning over to you for review, if all looks good, we can move this to QA.

asvinb commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js:
```diff
@@ -0,0 +1,89 @@
+		setUpdating( false );
+	}, [ fetchGoogleAdsAccountStatus ] );
+
+	useWindowFocusCallbackIntervalEffect( checkUpdatedAdsAccountStatus, 30 );
```
@ankitrox Even though the user has not clicked on the `Claim your Google Ads account` button, there'll be requests made to the API every 30s to check for the status, which is incorrect. The requests should be made only if the user clicked on the button. I suggest the following:
* Extract the `useWindowFocusCallbackIntervalEffect` logic in a new component. for e.g `ClaimAccountStatus` and then conditionally render `ClaimAccountStatus` if `updating` is `true`.

ankitrox replied:
As per our discussion with @joemcgill , let's keep it as is for now.

joemcgill replied:
For posterity, the reason this is ok is because the `ClaimAdsAccount` is only rendered when we've already determined that the account needs to be claimed. Since that claim can be completed externally to the component (e.g., via accepting the invite via an email link) we don't need to make this dependent on clicking the "claim" button.

asvinb replied:
👍

asvinb commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js:
```diff
@@ -0,0 +1,89 @@
+
+	return (
+		<Section.Card.Body className="gla-claim-ads-account-section">
+			<Flex gap={ 4 }>
```
@ankitrox We can avoid using the `Flex` and `FlexBlock` components here since most of the styles are scoped to the `gla-claim-ads-account-box` class name.

joemcgill reviewed (approved):
Thanks. Looks good to me.

asvinb commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.scss:
```diff
@@ -0,0 +1,30 @@
+		margin: 0;
+	}
+
+	.gla-claim-ads-account-box {
```
@ankitrox We can move `.gla-claim-ads-account-box` out of `.gla-claim-ads-account-section`. There is no need to have that level of speficity.

asvinb commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.scss:
```diff
@@ -0,0 +1,30 @@
+.gla-claim-ads-account-section {
+	padding-top: 0;
+
+	* {
```
@ankitrox I would remove that and instead simplify the CSS to the following:
```scss
.gla-claim-ads-account-box {
	border: 1px solid $gray-300;
	font-size: $gla-font-small;
	padding: $grid-unit-30;

	> * + * {
		margin: $grid-unit-10 0 0;
	}

	h4 {
		font-size: $gla-font-small;
		margin-top: 0;
	}

	p:nth-of-type(2) {
		color: $gray-700;
		font-style: italic;
	}

	button,
	.gla-loading-label {
		display: flex;
		margin-left: auto;
	}
}
```

ankitrox replied:
Thanks @asvinb .

We still need

```css
.gla-claim-ads-account-section {
	padding-top: 0;
}
```

to reduce spacing between card and account details.

asvinb replied:
Yes, we'll need the padding. Forgot to include that in my snippet.

asvinb commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.scss:
```diff
@@ -0,0 +1,30 @@
+			font-size: $gla-font-small;
+		}
+
+		p:nth-of-type(2) {
```
@ankitrox This is too brittle. Imagine if we add a new paragraph in the second position, the styles might be broken. I suggest we instead add a custom class name to the paragraph and style it accordingly instead of styling it based on its position.

ankitrox replied:
Thanks @asvinb 

Added a class `gla-ads-post-claim-instructions`

asvinb commented on js/src/components/google-combo-account-card/claim-ads-account/index.js:
```diff
@@ -0,0 +1,5 @@
+/**
```
@ankitrox There are not internal dependencies.

asvinb commented on js/src/components/google-combo-account-card/conversion-measurement-notice/index.js:
```diff
@@ -0,0 +1,26 @@
+ */
+import AppNotice from '.~/components/app-notice';
+
+const ConversionMeasurementNotice = () => {
```
@ankitrox Since we just need to render the `AppNotice` component, let's render it directly in `ConnectedAccountsActions` instead of having a new `ConversionMeasurementNotice` component.

ankitrox replied:
@asvinb IMO, I think we still keep `ConnectedAccountsActions`  and `ConversionMeasurementNotice`  components separate as `ConnectedAccountsActions`  component will wrap the business logic for all components that may need (or may not need) to show depending on certain conditions and then the components themselves (like `ConversionMeasurementNotice`) should only be UI focused giving them better decoupled behaviour and testability.

CC: @joemcgill

ankitrox replied:
@asvinb After our discussion, I have removed `ConversionMeasurementNotice` component and used `AppNotice` directly within `ConnectedAccountsActions`.

asvinb replied:
👍

ankitrox committed:
CR feedback.

ankitrox committed:
Remove ConversionMeasurementNotice component.

joemcgill commented:
To make it easier to test this PR, I've been using the following snippet in my mu-plugin to mock the API responses when an Ads account is unclaimed:

```php
add_filter( 'rest_post_dispatch', 'gla_mock_ads_unclaimed', 10, 3 );
function gla_mock_ads_unclaimed( $response, $server, $request ) {
	$method = $request->get_method();

	if ( '/wc/gla/ads/connection' === $request->get_route() ) {
		if ( 'GET' === $method ) {
			$data = [
				'id'       => 123456789,
				'currency' => 'USD',
				'status'   => 'incomplete',
				'step'     => 'link_merchant',
				'symbol'   => '$',
			];

			$response->set_data( $data );
		}
	}

	if ( '/wc/gla/ads/account-status' === $request->get_route() ) {
		$data = [
			'has_access'  => false,
			'step'        => 'link_merchant',
			'invite_link' => '',
		];

		$response->set_data( $data );
	}

	return $response;
}
```

ankitguptaindia commented:
Thanks for the PR @ankitrox I noticed when user claiming the the account, `Ads Conversation Measurement has been setup` message doesn't appearing in Google card. 

Please have a look in this screencast for my steps: 

https://github.com/user-attachments/assets/7ab2859a-c406-4f67-b2cb-21f10600a0e8

ankitrox committed:
Await till conversion measurement is setup.

ankitrox committed:
Do not render the claim card conditionally.

It prevents conversion action from being setup.

ankitrox commented:
> Thanks for the PR @ankitrox I noticed when user claiming the the account, `Ads Conversation Measurement has been setup` message doesn't appearing in Google card.
> 
> Please have a look in this screencast for my steps:
> 
>  Recording.932.mp4

This has been fixed with the latest changes.

![Screenshot 2024-10-29 at 6 58 54 PM](https://github.com/user-attachments/assets/633d9ade-564d-44aa-ac13-1f38d1fa7f76)

joemcgill commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js:
```diff
@@ -41,47 +42,48 @@ const ClaimAdsAccount = () => {
 		}
 	}, [ hasAccess, step, upsertAdsAccount ] );
 
+	const shouldClaimGoogleAdsAccount = Boolean(
```
If we're moving this from `ConnectedAccountsActions` to inside this component, then we also need to update the `checkUpdatedAdsAccountStatus()` callback so that is stops calling the `fetchGoogleAdsAccountStatus()` whenever `shouldClaimGoogleAdsAccount` is `false`. Otherwise, the API polling will continue even after the account is claimed.

ankitrox replied:
Made this change and pushed.

ankitrox committed:
Prevent continous polling.

ankitrox committed:
Add bottom padding for conversion measurement notice.

ankitrox committed:
Add bottom padding for conversion measurement notice.

ankitrox committed:
Reduce padding bottom.

ankitrox committed:
Revert redundant code.

ankitrox committed:
Remove redundant files.

joemcgill reviewed (approved):
The updates look good to me. Thank, @ankitrox

joemcgill commented:
@eason9487 this still needs to go through QA, but feel free to let us know if you have any structural concerns in the meantime.

ankitguptaindia commented:
Reported issue here https://github.com/woocommerce/google-listings-and-ads/pull/2644#issuecomment-2443684816 has been fixed and this PR is ready for next step.  Claim new account flow working as described in AC.  ✅ 

Quick Demo: 

https://github.com/user-attachments/assets/cad82518-4452-434f-8cf4-5520a2927ce9

asvinb committed:
Merge branch 'feature/2509-consolidate-google-account-cards' into update/2582-claim-ads-account

asvinb committed:
Port over changes.

joemcgill committed:
Improve 'Connect Google account' E2E tests

This eliminates some unnecessary page.goto() calls to speed up the tests and reduce some inconsistencies.

joemcgill commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js:
```diff
@@ -0,0 +1,86 @@
+
+	useWindowFocusCallbackIntervalEffect( checkUpdatedAdsAccountStatus, 30 );
+
+	const handleOnClick = () => {
```
One observation when testing this is that if someone clicks the Claim... button but then does not accept the invite, it might be nice to reset this value on the next `checkUpdatedAdsAccountStatus()` check so they could click the button to open the modal again. Otherwise, the have to refresh the page to get the button to work again.

asvinb replied:
@joemcgill I think that'll be confusing for the user 🤔 Once the user clicked the claim button, he will see the "Updating..." text for the button. Then shortly he will see the claim button back. The user might think there's an issue and will want to click the button again. Am not sure, I just feel it'll be weird for the user.

joemcgill replied:
Let's leave it for now.

joemcgill commented on js/src/components/google-ads-account-card/claim-account-button.js:
```diff
@@ -48,6 +53,7 @@ const ClaimAccountButton = ( { onClick = noop, ...restProps } ) => {
 			eventName="gla_open_ads_account_claim_invitation_button_click"
 			eventProps={ getEventProps() }
 			onClick={ handleClaimAccountClick }
+			loading={ ! inviteLink || loading }
```
Why is checking the `inviteLink` now necessary?

asvinb replied:
So when testing locally, sometimes i would see the API response (`/ads/account-status`) would take some time and we are displaying the button without actually having a link. This resulted in an empty popup several times because the data was still being resolved.

ankitrox replied:
@asvinb I think `loading` should be `true` in such cases where the data resolution is not completed yet.
CC: @joemcgill

joemcgill replied:
In that case, this component should probably check the `hasFinishedResolution` prop from `useGoogleAdsAccountStatus` and return `null` or a `Spinner` until we have resolved the address. That said, when I try to arbitrarily delay the account-status API response, this whole component is not showing, so I'm unsure how to reproduce what you experienced.

joemcgill commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js:
```diff
@@ -0,0 +1,86 @@
+					loading={ updating }
+					text={
+						updating
+							? __( 'Updating…', 'google-listings-and-ads' )
```
I'm not sure "Updating..." is the best copy for this situation. In the current UI, we show "Updating..." instead of the "Claim..." button based on the loading prop returned from the `useUpsertAdsAccount()` hook. If we want to leave the button in a "pending" state while we're polling for updates to the account-status endpoint, we could use a word like "Waiting..." or "Checking access...". What do you think?

asvinb replied:
Sure, I've updated the wording to "Waiting..."

joemcgill reviewed (commented):
Left a few observations/questions on the latest changes, but this seems to be working as intended.

asvinb committed:
Rename label.

asvinb committed:
Show conversion notice when account is being updated.

asvinb committed:
Add callback.

joemcgill reviewed (changes requested):
This is looking better, although it looks like the new logic for only displaying the conversion action notice when it is first completed is not yet working as expected.

Right now, after the account is claimed an the `upsertAdsAccount()` function is called. The notice is shown without validating that the conversion action was actually completed.

joemcgill committed:
Simplify loading for ClaimAccountsButton

joemcgill committed:
Use useEffect to toggleconversion notice state

joemcgill committed:
Fix jest tests

joemcgill reviewed (approved):
@asvinb I've pushed a couple of commits that address the issues I noticed. I'm going to self approve, but please take a look and let me know what you think.

asvinb committed:
Update E2E tests.

asvinb commented:
Thanks @joemcgill . Changes look good. I've also updated the E2E tests 👍

asvinb committed:
Revert changes.

joemcgill commented on tests/e2e/specs/setup-mc/step-1-accounts.test.js:
```diff
@@ -397,26 +410,15 @@ test.describe( 'Set up accounts', () => {
-
-				await setUpAccountsPage.goto();
-			} );
+		test( 'should see the merchant center id once account has been claimed', async () => {
```
This will pass, but doesn't really test that the claim was successful because the MC ID is shown even when the claim account message is visible. I think a better test would wait for the UI to update and confirm that after we receive a successful claim response the UI looks as intended.

Here's an idea that seems to work, though I'm not sure the focus action is working as expected because it take a long time for the account-status polling function to kick in:


```diff
diff --git a/tests/e2e/specs/setup-mc/step-1-accounts.test.js b/tests/e2e/specs/setup-mc/step-1-accounts.test.js
index 641bfce68..51cc5fbed 100644
--- a/tests/e2e/specs/setup-mc/step-1-accounts.test.js
+++ b/tests/e2e/specs/setup-mc/step-1-accounts.test.js
@@ -408,17 +408,34 @@ test.describe( 'Set up accounts', () => {
 			await popupPage.waitForLoadState();
 			const url = popupPage.url();
 			expect( url ).toMatch( /^https:\/\/example\.com(\/|\?|$)/ );
+
+			await popupPage.close();
 		} );
 
 		test( 'should see the merchant center id once account has been claimed', async () => {
 			await setUpAccountsPage.mockAdsStatusClaimed();
 			await setUpAccountsPage.mockAdsAccountConnected();
 
-			await setUpAccountsPage.goto();
+			const statusResponse =
+				setUpAccountsPage.registerAdsAccountStatusResponse();
+
+			await setUpAccountsPage.page.dispatchEvent( 'body', 'focus' );
+
+			await statusResponse;
+
 			const googleAccountCard = setUpAccountsPage.getGoogleAccountCard();
 			await expect( googleAccountCard ).toContainText(
 				'Google Ads ID: 12345'
 			);
+
+			await expect(
+				googleAccountCard.getByText(
+					'Google Ads conversion measurement has been set up for your store.',
+					{
+						exact: true,
+					}
+				)
+			).toBeVisible();
 		} );
 	} );
 
diff --git a/tests/e2e/utils/pages/setup-mc/step-1-set-up-accounts.js b/tests/e2e/utils/pages/setup-mc/step-1-set-up-accounts.js
index cf50408e4..be68be31f 100644
--- a/tests/e2e/utils/pages/setup-mc/step-1-set-up-accounts.js
+++ b/tests/e2e/utils/pages/setup-mc/step-1-set-up-accounts.js
@@ -366,6 +366,19 @@ export default class SetUpAccountsPage extends MockRequests {
 		);
 	}
 
+	/**
+	 * Register the responses when checking for account-status.
+	 *
+	 * @return {Promise<import('@playwright/test').Response>} The response.
+	 */
+	registerAdsAccountStatusResponse() {
+		return this.page.waitForResponse(
+			( response ) =>
+				response.url().includes( '/gla/ads/account-status' ) &&
+				response.status() === 200
+		);
+	}
+
 	/**
 	 * Get terms checkbox.
 	 *
```

asvinb replied:
Thanks @joemcgill . I've updated the test. Can you kindly take a look now?

asvinb committed:
Better E2E tests.

asvinb committed:
Merge branch 'feature/2509-consolidate-google-account-cards' into update/2582-claim-ads-account

joemcgill committed:
E2E: Remove conversion action case from Ads claim test

joemcgill committed:
E2E: Remove unused fixture method

joemcgill reviewed (approved):
The latest updates here are looking good to me.

I tried looking for a better way handle the Claim Ads E2E test so that the successful claim could be checked without a page refresh, but waiting for the `useWindowFocusCallbackIntervalEffect` to run makes that test take too long, so I modified it again to just confirm that the AccountCard shows the 'Connected' message in [3e2d8ee](https://github.com/woocommerce/google-listings-and-ads/pull/2644/commits/3e2d8ee098428079499c4239955261cd13c6c086) with additional cleanup in [0f38c5e](https://github.com/woocommerce/google-listings-and-ads/pull/2644/commits/0f38c5e3bf3087f78d44b92173d25846e008e812).

I also cherry-picked the changes from [80764f0](https://github.com/woocommerce/google-listings-and-ads/pull/2644/commits/80764f0633c9bdfc64f43900c4b8d197909af939) that fixes some flakiness with the account creation tests.

eason9487 commented on js/src/components/google-combo-account-card/connected-google-combo-account-card.js:
```diff
@@ -34,13 +50,53 @@ const ConnectedGoogleComboAccountCard = () => {
+		if ( initConnected.current === null ) {
+			initConnected.current = isConnected;
+		}
+	}, [ initConnected, isConnected ] );
```
I don't see a requirement explicitly asking the conversion measurement notice to be shown only once after an account is transformed from unconnected to connected. In addition, this logic is inconsistent with another logic that also controls the rendering of conversion measurement notice.

https://github.com/woocommerce/google-listings-and-ads/blob/0f38c5e3bf3087f78d44b92173d25846e008e812/js/src/components/google-ads-account-card/google-ads-account-card.js#L63-L75

Is this intentionally inconsistent? Or is the additional logic here unnecessary?

joemcgill replied:
Good observation! There is not an explicit requirement for this, but the design doc implies that after all accounts finish connecting and the combo card collapses, that this would no longer be shown. Additionally, only showing this whenever the `conversion_action` setup step has been completed more closely aligns the notice to its purpose.

The difference in logic is mainly because we didn't update the `GoogleAdsAccountCard` to make use of the new `useGoogleAdsAccountReady()` hook but we certainly could. 

Functionally the logic for both are nearly the same.

The new logic requires: 
- we have a connection (could be 'connected' or 'incomplete')
- we have access (e.g., the invite has been accepted)
- and the setup step is past the 'conversion_action' step (which includes  'link_merchant', 'billing', or '' if setup is complete)

The old logic requires:
- we have a connection (could be 'connected' or 'incomplete' if the step is 'link_merchant')
- we have access (otherwise we never reach this logic and the `NonConnected` component is returned)
- we have a setup step that is either 'link_merchant' or '' (when connected). If the step is 'conversion_action' the `NonConnected` component is returned.

The only case in the old component where the action will NOT be shown is if an account is set up resulting in the 'billing' step being returned. I've confirmed this by creating a new account during the secondary Create Ads setup flow, which I believe to be an oversight.

Updating the conversion measurement logic on the `GoogleAdsAccountCard` is out of scope for this PR, but it would be pretty safe to swap in the new hook if you'd prefer. Otherwise, I'd suggest leaving the `GoogleAdsAccountCard` logic as is and updating the logic in a follow-up issue.

eason9487 replied:
> There is not an explicit requirement for this, but the design doc implies that after all accounts finish connecting and the combo card collapses, that this would no longer be shown.

![1](https://github.com/user-attachments/assets/ece67612-d545-41e4-a736-7320d295b837)

- Only the paths with blue arrow hide it.
- The remaining paths with red arrow don't hide it.

The design document doesn't appear to imply this behavior. Instead, the design document is more like the logic that is not defined. I'm not so sure that's a valid reason to support the need of this logic.

> Additionally, only showing this whenever the `conversion_action` setup step has been completed more closely aligns the notice to its purpose.

I would have thought it was a good thing to let merchants always see the notice whenever they are viewing the connected Google Ads account card.

---

If it's not so necessary, I would suggest removing the logic that determines whether the initial state is transformed from unconnected to connected. But if it's indeed necessary, the current implementation doesn't work. With #2660, there are three problems:

### 1. The notice is still shown after disconnected

https://github.com/user-attachments/assets/4c563157-aefe-4e37-899c-d6a2388972f7

### 2. When starting with connected state, the notice is not shown after disconnected and reconnected

https://github.com/user-attachments/assets/d454c3ad-0114-4ec7-a4da-8975f73a1931

### 3. The notice is still shown after the combo card collapses, which doesn't work as stated

> [...], but the design doc implies that after all accounts finish connecting and the combo card collapses, that this would no longer be shown.

https://github.com/user-attachments/assets/1a7eb1f1-a19e-46c6-8ecd-8ff2e6328b4d

asvinb replied:
@eason9487 This has been fixed and updated. Also I think it's safe to say that whenever the "Cancel" button is clicked, we can hide the notice.

eason9487 commented on js/src/components/google-combo-account-card/connected-google-combo-account-card.js:
```diff
@@ -34,13 +50,53 @@ const ConnectedGoogleComboAccountCard = () => {
+		if ( initConnected.current === null ) {
+			initConnected.current = isConnected;
+		}
+	}, [ initConnected, isConnected ] );
```
The `initConnected` returned from `useRef` doesn't need to be a part of `useEffect` deps.

asvinb replied:
👍

eason9487 commented on js/src/components/google-combo-account-card/connected-google-combo-account-card.js:
```diff
@@ -1,9 +1,16 @@
 import AccountCard, { APPEARANCE } from '../account-card';
 import ConnectAds from './connect-ads';
 import AccountDetails from './account-details';
+import ConnectedAdsAccountsActions from './connected-ads-account-actions';
```
The "Account**s**" part doesn't follow  naming convention.

asvinb replied:
👍

eason9487 commented on js/src/components/google-combo-account-card/connected-ads-account-actions.js:
```diff
@@ -0,0 +1,47 @@
+ * @param {boolean} props.showConversionMeasurementNotice Whether to show the conversion measurement notice.
+ * @return {JSX.Element} Connected accounts actions.
+ */
+const ConnectedAdsAccountsActions = ( {
```
Suggest replacing "Accounts" with "Account" to be consistent with the file name.

asvinb replied:
👍 Component has been renamed `ConnectedAdsAccountDetail`

eason9487 commented on js/src/components/google-combo-account-card/connected-google-combo-account-card.js:
```diff
@@ -59,10 +122,15 @@ const ConnectedGoogleComboAccountCard = () => {
+						showConversionMeasurementNotice
+					}
+					claimGoogleAdsAccount={ shouldClaimGoogleAdsAccount }
+				/>
```
Suggest rendering this block via the `detail` prop with setting the `expandedDetail` prop can reduce the needs of layout styles. Furthermore, the names of `ConnectedAdsAccountsActions` and its file could be renamed to something like `ConnectedAdsAccountDetail` as the Google Ads conversion measurement notice is not quite considered an action.

For example:

https://github.com/woocommerce/google-listings-and-ads/blob/0f38c5e3bf3087f78d44b92173d25846e008e812/js/src/components/google-combo-account-card/connected-google-combo-account-card.scss#L7-L18

https://github.com/woocommerce/google-listings-and-ads/blob/0f38c5e3bf3087f78d44b92173d25846e008e812/js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js#L48

https://github.com/woocommerce/google-listings-and-ads/blob/0f38c5e3bf3087f78d44b92173d25846e008e812/js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.scss#L1-L3

asvinb replied:
Sounds good. Updated @eason9487

eason9487 commented on js/src/components/google-combo-account-card/connected-ads-account-actions.js:
```diff
@@ -0,0 +1,47 @@
+import ClaimAdsAccount from './claim-ads-account';
+
+/**
+ * Renders the actionable components for connected accounts.
```
Not all components rendered by it are actionable.

asvinb replied:
Updated comment.

eason9487 commented on js/src/components/google-combo-account-card/connected-ads-account-actions.js:
```diff
@@ -0,0 +1,47 @@
+			{ claimGoogleAdsAccount && <ClaimAdsAccount /> }
+
+			{ showConversionMeasurementNotice && (
+				<AppNotice
```
Using the `Notice` imported from `'@wordpress/components'` should be enough as all `AppNotice` styles are overridden.

![image](https://github.com/user-attachments/assets/2d38f077-16a6-4579-a630-7e297be96583)

asvinb replied:
👍

eason9487 commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.scss:
```diff
@@ -0,0 +1,29 @@
+
+.gla-claim-ads-account-box {
+	border: 1px solid $gray-300;
+	font-size: $gla-font-small;
```
Setting font size is unnecessary as all children have been set their font size.

asvinb replied:
👍

eason9487 commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.scss:
```diff
@@ -0,0 +1,29 @@
+	font-size: $gla-font-small;
+	padding: $grid-unit-30;
+
+	> * + * {
```
Prefer to set a CSS class name on components or elements that need (re)setting spacing rather than using `*`.

asvinb replied:
👍 . I used flex on the parent container and set the gap for the spacing.

eason9487 commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.scss:
```diff
@@ -0,0 +1,29 @@
+	}
+
+	button,
+	.gla-loading-label {
```
Will the `ClaimAdsAccount` component have any elements that use `.gla-loading-label`?

asvinb replied:
Nope. Removed.

eason9487 commented on js/src/components/google-ads-account-card/claim-account-button.js:
```diff
@@ -27,10 +27,15 @@ import useEventPropertiesFilter from '.~/hooks/useEventPropertiesFilter';
-	const { inviteLink } = useGoogleAdsAccountStatus();
+const ClaimAccountButton = ( {
+	onClick = noop,
+	loading = false,
```
The original `restProps` also forwards the `loading` to `AppButton`. Destructuring and passing it may not be necessary.

asvinb replied:
👍

eason9487 commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js:
```diff
@@ -0,0 +1,86 @@
+ */
+const ClaimAdsAccount = () => {
+	const [ updating, setUpdating ] = useState( false );
+	const { googleAdsAccount } = useGoogleAdsAccount();
+	const { hasAccess } = useGoogleAdsAccountStatus();
+	const { fetchGoogleAdsAccountStatus } = useAppDispatch();
+
+	const shouldClaimGoogleAdsAccount = Boolean(
+		googleAdsAccount?.id && hasAccess === false
+	);
+
+	const checkUpdatedAdsAccountStatus = useCallback( async () => {
+		if ( ! shouldClaimGoogleAdsAccount ) {
+			return;
+		}
+
+		await fetchGoogleAdsAccountStatus();
+	}, [ fetchGoogleAdsAccountStatus, shouldClaimGoogleAdsAccount ] );
+
+	useWindowFocusCallbackIntervalEffect( checkUpdatedAdsAccountStatus, 30 );
```
https://github.com/woocommerce/google-listings-and-ads/blob/0f38c5e3bf3087f78d44b92173d25846e008e812/js/src/components/google-combo-account-card/connected-google-combo-account-card.js#L96-L98

https://github.com/woocommerce/google-listings-and-ads/blob/0f38c5e3bf3087f78d44b92173d25846e008e812/js/src/components/google-combo-account-card/connected-google-combo-account-card.js#L127-L132

Considering the upper layer has the same check and controls the rendering of this component, maybe using `useWindowFocusCallbackIntervalEffect( fetchGoogleAdsAccountStatus, 30 );` here would be sufficient.

asvinb replied:
Thanks for catching that. Updated.

eason9487 reviewed (commented):
Found an issue that was not introduced by this PR but must be raised.

If a merchant doesn't authorize all required scopes when connecting Google account, the automatically created process may get stuck and can not continue. Refreshing webpage also remains in the same stuck state.

The following video demos the issue when the Google Ads scope (`'https://www.googleapis.com/auth/adwords'`) is not selected.

https://github.com/user-attachments/assets/285b3d03-b21d-4ea5-9742-b513a755c7d5

The original `scope.glaRequired` checking logic is no longer applicable to the current implementation. The conditions below must check both `scope.gmcRequired` and `scope.adsRequired`.

https://github.com/woocommerce/google-listings-and-ads/blob/0f38c5e3bf3087f78d44b92173d25846e008e812/js/src/components/google-combo-account-card/index.js#L21-L25

asvinb committed:
Address CR feedback.

asvinb committed:
Use detail prop.

asvinb committed:
Match spacing with designs.

joemcgill commented:
@eason9487 I've [left some feedback](https://github.com/woocommerce/google-listings-and-ads/pull/2644#discussion_r1833445209) about one of the questions you raised, but I think it's ready otherwise.

eason9487 commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js:
```diff
@@ -0,0 +1,74 @@
+	};
+
+	return (
+		<Section.Card.Body className="gla-claim-ads-account-section">
```
It no longer needs `Section.Card.Body` and `.gla-claim-ads-account-section` CSS style.

https://github.com/woocommerce/google-listings-and-ads/blob/1eadc097b0aa1c5193658426dc6f8d634fbc90fd/js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.scss#L1-L2

asvinb replied:
👍

eason9487 commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js:
```diff
@@ -0,0 +1,74 @@
+
+	const checkUpdatedAdsAccountStatus = useCallback( async () => {
+		await fetchGoogleAdsAccountStatus();
+	}, [ fetchGoogleAdsAccountStatus ] );
```
Wrapping `fetchGoogleAdsAccountStatus` with `useCallback` is redundant.

asvinb replied:
👍

eason9487 commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js:
```diff
@@ -0,0 +1,74 @@
+						updating
+							? __( 'Waiting…', 'google-listings-and-ads' )
+							: __(
+									'Claim your Google Ads account',
```
This wording doesn't match the one in requirement or design document.

![image](https://github.com/user-attachments/assets/45bfd09c-b10b-4c5f-9ed7-ff627fa1daab)

asvinb replied:
Thanks! Updated!

eason9487 commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.scss:
```diff
@@ -0,0 +1,30 @@
+
+		.app-button {
+			display: flex;
+			margin-left: auto;
```
After the parent was changed to flex layout, it could be aligned by:

```suggestion
			align-self: flex-end;
```

asvinb replied:
Nice catch! Updated!

eason9487 commented on js/src/components/google-combo-account-card/connected-ads-account-detail.scss:
```diff
@@ -0,0 +1,7 @@
+.gla-connected-ads-account-detail {
+	margin-top: $grid-unit-50 - $grid-unit-15;
```
This spacing should not be set.

Design document:
![image](https://github.com/user-attachments/assets/487d21d8-fff2-4d10-a00c-1baf5938613f)

![image](https://github.com/user-attachments/assets/ae4851a0-471f-4a7a-9f55-61db5ecd4a9d)

Implementation:
![image](https://github.com/user-attachments/assets/3d8e57dc-f94d-4423-b150-10230b8f94e0)

![image](https://github.com/user-attachments/assets/7c2299b6-a378-4746-8096-8b82e1eeab25)

asvinb replied:
👍

eason9487 commented on js/src/components/google-combo-account-card/index.js:
```diff
@@ -18,7 +18,7 @@ export default function GoogleComboAccountCard( { disabled = false } ) {
 	const isConnected = google?.active === 'yes';
 
-	if ( isConnected && scope.glaRequired ) {
+	if ( isConnected && scope.gmcRequired && scope.glaRequired ) {
```
Continue on https://github.com/woocommerce/google-listings-and-ads/pull/2644#pullrequestreview-2420088853.

> The original `scope.glaRequired` checking logic is no longer applicable to the current implementation. The conditions below must check both `scope.gmcRequired` and `scope.adsRequired`.

The `scope.glaRequired` is still used, so the problem is not addressed.

https://github.com/user-attachments/assets/9777ec7e-b91d-4cc6-ba97-7993a40d4315

In addition, [this condition](https://github.com/woocommerce/google-listings-and-ads/blob/1eadc097b0aa1c5193658426dc6f8d634fbc90fd/js/src/components/google-combo-account-card/index.js#L25) needs an adjustment as well in order to request full access.

![image](https://github.com/user-attachments/assets/a785ce96-3997-4a5d-bad8-f84c91567153)

asvinb replied:
@eason9487 I've updated the conditions. Can you double check please?

asvinb committed:
Revert logic to display conversion notice.

eason9487 commented on tests/e2e/specs/setup-mc/step-1-accounts.test.js:
```diff
@@ -149,12 +149,9 @@ test.describe( 'Set up accounts', () => {
-			await setUpAccountsPage.mockJetpackConnected(
-				'Test user',
-				'jetpack@example.com'
-			);
+			await setUpAccountsPage.mockJetpackNotConnected();
```
The code comment is incorrect.

asvinb replied:
Updated.

eason9487 commented on tests/e2e/specs/setup-mc/step-1-accounts.test.js:
```diff
@@ -165,7 +162,23 @@ test.describe( 'Set up accounts', () => {
+		} );
+
 		test( 'should see their WPORG email, "Google" title & connect button', async () => {
+			// Mock Jetpack as disconnected
+			await setUpAccountsPage.mockJetpackConnected();
```
The code comment is incorrect.

asvinb replied:
Updated.

eason9487 commented on tests/e2e/specs/setup-mc/step-1-accounts.test.js:
```diff
@@ -616,6 +613,61 @@ test.describe( 'Set up accounts', () => {
+			await setUpAccountsPage.mockGoogleConnected();
+			await setUpAccountsPage.mockMCHasAccounts();
+			await setUpAccountsPage.mockMCConnected();
+			await setUpAccountsPage.fulfillAdsAccounts( { id: 12345 } );
```
The payload should be `Array<Object>` type.

asvinb replied:
👍

eason9487 commented on tests/e2e/specs/setup-mc/step-1-accounts.test.js:
```diff
@@ -616,6 +613,61 @@ test.describe( 'Set up accounts', () => {
+
+			const [ popupPage ] = await Promise.all( [
+				page.waitForEvent( 'popup' ),
+				await googleAccountCard
```
The `await` is unnecessary.

asvinb replied:
👍

eason9487 commented on tests/e2e/specs/setup-mc/step-1-accounts.test.js:
```diff
@@ -616,6 +613,61 @@ test.describe( 'Set up accounts', () => {
+
+			await popupPage.waitForLoadState();
+			const url = popupPage.url();
+			expect( url ).toMatch( /^https:\/\/example\.com(\/|\?|$)/ );
```
```suggestion
			expect( url ).toMatch( /^https:\/\/example\.com\/?$/ );
```

asvinb replied:
👍

asvinb committed:
Address CR feedback.

asvinb committed:
Update tests.

asvinb committed:
Fix rendering issue.

eason9487 commented on js/src/components/google-ads-account-card/claim-account-button.js:
```diff
@@ -42,6 +42,10 @@ const ClaimAccountButton = ( { onClick = noop, ...restProps } ) => {
 		onClick( event );
 	};
 
+	if ( ! hasFinishedResolution ) {
+		return null;
+	}
```
While viewing #2651, discovered that if clicking fast enough, it may open a blank page. Checking `hasFinishedResolution` can not ensure that the `inviteLink` is already available.

https://github.com/user-attachments/assets/79ee9cbf-8006-4d04-81d6-1ca39b439004

Also the claim button will become continually locked on "Waiting..." and it will not be able to open the pop-up window via the button again unless refresh the webpage.

https://github.com/woocommerce/google-listings-and-ads/blob/972f45238d9a780563ee35e1becce4f6d2573222/js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js#L49-L53

asvinb replied:
@eason9487 I've added the loading indicator if `inviteLink` is `null`. It should fix the issue you noted.

joemcgill replied:
This approach was replaced, based on [this observation](https://github.com/woocommerce/google-listings-and-ads/pull/2644#discussion_r1835815402).

joemcgill replied:
> Also the claim button will become continually locked on "Waiting..." and it will not be able to open the pop-up window via the button again unless refresh the webpage.

I've addressed this in [d074504](https://github.com/woocommerce/google-listings-and-ads/pull/2651/commits/d074504253e2c10c91bf944f062bcd51efb80727).

eason9487 commented on js/src/components/google-combo-account-card/connected-google-combo-account-card.js:
```diff
@@ -51,6 +86,18 @@ const ConnectedGoogleComboAccountCard = () => {
+		( Boolean( creatingWhich ) && ! shouldClaimGoogleAdsAccount ) ||
+		( ! showConnectAds && finalizeAdsAccountCreation );
+
+	const showConversionMeasurementNotice =
+		( googleAdsAccount?.status === GOOGLE_ADS_ACCOUNT_STATUS.CONNECTED ||
+			googleAdsAccount?.step === 'link_merchant' ) &&
```
As mentioned in https://github.com/woocommerce/google-listings-and-ads/pull/2644#discussion_r1833445209:

> The only case in the old component where the action will NOT be shown is if an account is set up resulting in the 'billing' step being returned. I've confirmed this by creating a new account during the secondary Create Ads setup flow, which I believe to be an oversight.

Perhaps this condition should be taken into account when `step` is `'billing'`.

https://github.com/user-attachments/assets/4b284b1a-eabf-455a-831e-27ca56afce07

joemcgill replied:
I'm not sure I fully understand the suggestion here, but in order to account for both cases (when the step is either 'link_merchant' or 'billing', I've updated the logic used by both components in [ac3eef9](https://github.com/woocommerce/google-listings-and-ads/pull/2644/commits/ac3eef9d1113683baa76f27671c9508ddec20486) to use a common utility function that includes the 'billing' case.

asvinb committed:
Add loading indicator while invite link is being resolved.

joemcgill commented on js/src/components/google-combo-account-card/connected-google-combo-account-card.js:
```diff
@@ -51,6 +86,18 @@ const ConnectedGoogleComboAccountCard = () => {
+		( Boolean( creatingWhich ) && ! shouldClaimGoogleAdsAccount ) ||
+		( ! showConnectAds && finalizeAdsAccountCreation );
+
+	const showConversionMeasurementNotice =
+		( googleAdsAccount?.status === GOOGLE_ADS_ACCOUNT_STATUS.CONNECTED ||
+			googleAdsAccount?.step === 'billing' ) &&
+		! shouldClaimGoogleAdsAccount;
```
Happy for us to remove the `useEffect` logic that only shows this when the conversion action has first been set up, but if we're using similar logic to the `GoogleAdsAccountCard` we should make it consistent, i.e.

```suggestion
	const showConversionMeasurementNotice =
		( googleAdsAccount?.status === GOOGLE_ADS_ACCOUNT_STATUS.CONNECTED ||
			googleAdsAccount?.step === 'link_merchant' ) &&
		! shouldClaimGoogleAdsAccount;
```

Also, thought it doesn't hurt things, I don't think the `! shouldClaimGoogleAdsAccount` is necessary, because in neither condition above would `shouldClaimGoogleAdsAccount` resolve to `true`.

joemcgill replied:
I've handled this in [ac3eef9](https://github.com/woocommerce/google-listings-and-ads/pull/2644/commits/ac3eef9d1113683baa76f27671c9508ddec20486) and moved this logic to a utility function to avoid inconsistent logic.

joemcgill commented on js/src/components/google-ads-account-card/claim-account-button.js:
```diff
@@ -41,10 +42,10 @@ const ClaimAccountButton = ( { onClick = noop, ...restProps } ) => {
 	return (
 		<AppButton
 			{ ...restProps }
+			loading={ loading || ! inviteLink }
```
Using the `inviteLink` value as part of the loading value is problematic, because it is possible for the `inviteLink` to be an empty string, (ie. `inviteLink: ""`). The case of the `inviteLink` being empty seems to be related to a strange race condition.

When no account is connected, the initial state for account-status will be:

```json
{
    "has_access": false,
    "step": "",
    "invite_link": ""
}
```

To create a new account, we call `upsertAdsAccount()`, which internally makes a direct POST request (bypassing actions altogether), followed by refetching the Ads account and the new Ads account status.

This is where a race condition can happen because the store will still have the initial account-status with an `invite_link: ""` and `has_access: false` but the Ads account will be resolved, causing this logic to be true:

```js
const shouldClaimGoogleAdsAccount = Boolean(
	googleAdsAccount?.id && hasAccess === false
);
```

I think the safest thing to do is to either invalidate the account-status data whenever we're creating a new account, or add a condition to `shouldClaimGoogleAdsAccount` that waits for the `upsertAdsAccount()` `loading` value to be `false`. Here's an example:

```diff
diff --git a/js/src/components/google-ads-account-card/claim-account-button.js b/js/src/components/google-ads-account-card/claim-account-button.js
index e9afa0a78..b53ded9c9 100644
--- a/js/src/components/google-ads-account-card/claim-account-button.js
+++ b/js/src/components/google-ads-account-card/claim-account-button.js
@@ -45,7 +45,7 @@ const ClaimAccountButton = ( { onClick = noop, loading, ...restProps } ) => {
 	return (
 		<AppButton
 			{ ...restProps }
-			loading={ loading || ! inviteLink }
+			loading={ loading }
 			eventName="gla_open_ads_account_claim_invitation_button_click"
 			eventProps={ getEventProps() }
 			onClick={ handleClaimAccountClick }
diff --git a/js/src/components/google-combo-account-card/connected-google-combo-account-card.js b/js/src/components/google-combo-account-card/connected-google-combo-account-card.js
index ab2a4eab8..536db2912 100644
--- a/js/src/components/google-combo-account-card/connected-google-combo-account-card.js
+++ b/js/src/components/google-combo-account-card/connected-google-combo-account-card.js
@@ -48,7 +48,7 @@ const ConnectedGoogleComboAccountCard = () => {
 	const { invalidateResolution } = useAppDispatch();
 	const { googleAdsAccount } = useGoogleAdsAccount();
 	const { hasAccess, step } = useGoogleAdsAccountStatus();
-	const [ upsertAdsAccount, { action } ] = useUpsertAdsAccount();
+	const [ upsertAdsAccount, { action, loading } ] = useUpsertAdsAccount();
 
 	const finalizeAdsAccountCreation =
 		hasAccess === true && step === 'conversion_action';
@@ -73,7 +73,7 @@ const ConnectedGoogleComboAccountCard = () => {
 	// @TODO: edit mode implementation in 2605
 	const editMode = false;
 	const shouldClaimGoogleAdsAccount = Boolean(
-		googleAdsAccount?.id && hasAccess === false
+		! loading && googleAdsAccount?.id && hasAccess === false
 	);
 
 	const hasExistingGoogleMCAccounts = existingGoogleMCAccounts?.length > 0;
```

joemcgill replied:
Handled in [43c5ab9](https://github.com/woocommerce/google-listings-and-ads/pull/2644/commits/43c5ab9cfc4e29db6454d1462da52c134acf668e) and [6e2b00d](https://github.com/woocommerce/google-listings-and-ads/pull/2644/commits/6e2b00df921ef54087121787981d2c333815a959).

joemcgill reviewed (changes requested):
Happy to remove the logic to conditionally show the conversion notice, but I don't think the changes are quite right yet. Left two suggestions inline.

joemcgill committed:
Don't show Claim Ads while upsert in progress

joemcgill committed:
Don't render the claim button without a URL

joemcgill committed:
Consolodate logic for showing Ads Conversion notice

joemcgill committed:
Limit Ads claim updates to 3 retries

joemcgill commented:
@eason9487 this is ready for another review, though it may be easier to test everything if we go ahead and merge #2651 into this PR, since testing that the `inviteLink` issues are working properly require a new Ads account to be created via the UI, so let us know if you want to review these together.

A summary of the latest updates:

1. Based on [these observations](https://github.com/woocommerce/google-listings-and-ads/pull/2644#discussion_r1835815402) the logic for showing the claim component has been updated in [43c5ab9](https://github.com/woocommerce/google-listings-and-ads/pull/2644/commits/43c5ab9cfc4e29db6454d1462da52c134acf668e) so it's not shown until the accounts-status is refreshed.
2. Even with ☝🏻 addressed, it's technically possible for the API response to include an empty URL, so I've chosen not to show the button when there is no URL to send them to and only show the claim message instead in [6e2b00d](https://github.com/woocommerce/google-listings-and-ads/pull/2644/commits/6e2b00df921ef54087121787981d2c333815a959)
3. To keep the logic in sync for showing the Ads Conversion setup notice, I've created a new utility function in [ac3eef9](https://github.com/woocommerce/google-listings-and-ads/pull/2644/commits/ac3eef9d1113683baa76f27671c9508ddec20486)
4. To avoid the button getting continually stuck on "Waiting.." if the merchant closes the popup without claiming, I've added a retry limit in https://github.com/woocommerce/google-listings-and-ads/pull/2644#discussion_r1837281799, after which the button will be shown again.

Let me know your thoughts.

eason9487 commented on js/src/components/google-ads-account-card/claim-account-button.js:
```diff
@@ -42,9 +43,15 @@ const ClaimAccountButton = ( { onClick = noop, ...restProps } ) => {
 	return (
 		<AppButton
 			{ ...restProps }
+			loading={ loading }
```
Same as https://github.com/woocommerce/google-listings-and-ads/pull/2644#discussion_r1832501829

joemcgill replied:
[27fa0b1](https://github.com/woocommerce/google-listings-and-ads/pull/2644/commits/27fa0b14dd9f95c363fc965261cd7532bf90da67)

eason9487 commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js:
```diff
@@ -0,0 +1,91 @@
+ */
+const ClaimAdsAccount = () => {
+	const [ claiming, setClaiming ] = useState( false );
+	const retries = useRef( CLAIM_RETRIES );
```
It would be better to follow the naming convention to have a suffix `Ref`.

eason9487 commented on js/src/components/google-combo-account-card/claim-ads-account/claim-ads-account.js:
```diff
@@ -0,0 +1,91 @@
+				if ( retries.current < 1 ) {
+					setClaiming( false );
+				}
+			} )
```
In the worst-case scenario, it will take a merchant to stay on this page and wait for about 60 seconds before seeing the button return to enabled. The waiting time is obviously too long. Probably the merchant will reload the webpage before discovering the mechanism. Therefore, I'm not pretty sure if this retry is a good idea although it technically solves the concern.

It may be better to just leave the button always enabled and show the loading indicator beside it or not even show any loading/waiting state as the original Google Ads card does.

joemcgill replied:
For simplicity, I've removed all of the loading logic in [6c93d64](https://github.com/woocommerce/google-listings-and-ads/pull/2644/commits/6c93d643464e4697b6ff67b867b05e3d63b215c6) and the button will always be clickable and as long as the `ClaimAdsAccount` component is visible, it will always poll the API for updates on the account-status.

eason9487 reviewed (commented):
The retry mechanism is less than ideal, but if showing a waiting indicator is necessary, then the current implementation can be kept for now.

 The naming of `useRef` variable and redundant `loading` forwards still need to be resolved.

joemcgill committed:
Remove redundant loading prop from ClaimAccountButton

joemcgill committed:
Remove waiting state from ClaimAdsAccount component
